mongo/client_bulk_write.go

```diff
@@ -0,0 +1,718 @@
+			_, ok := batches.writeErrors[0]
+			hasSuccess = !ok
+		} else {
+			hasSuccess = len(batches.writeErrors) < len(bw.models)
+		}
+		if hasSuccess {
+			exception.PartialResult = batches.result
+		}
+		return *exception
+	}
+	return err
+}
+
+func (bw *clientBulkWrite) newCommand() func([]byte, description.SelectedServer) ([]byte, error) {
+	return func(dst []byte, desc description.SelectedServer) ([]byte, error) {
+		dst = bsoncore.AppendInt32Element(dst, "bulkWrite", 1)
+
+		dst = bsoncore.AppendBooleanElement(dst, "errorsOnly", bw.errorsOnly)
+		if bw.bypassDocumentValidation != nil && (desc.WireVersion != nil && driverutil.VersionRangeIncludes(*desc.WireVersion, 4)) {
+			dst = bsoncore.AppendBooleanElement(dst, "bypassDocumentValidation", *bw.bypassDocumentValidation)
+		}
+		if bw.comment != nil {
+			comment, err := marshalValue(bw.comment, bw.client.bsonOpts, bw.client.registry)
+			if err != nil {
+				return nil, err
+			}
+			dst = bsoncore.AppendValueElement(dst, "comment", comment)
+		}
+		dst = bsoncore.AppendBooleanElement(dst, "ordered", bw.ordered == nil || *bw.ordered)
+		if bw.let != nil {
+			let, err := marshal(bw.let, bw.client.bsonOpts, bw.client.registry)
+			if err != nil {
+				return nil, err
+			}
+			dst = bsoncore.AppendDocumentElement(dst, "let", let)
+		}
+		return dst, nil
+	}
+}
+
+type cursorInfo struct {
+	Ok        bool
+	Idx       int32
+	Code      *int32
+	Errmsg    *string
+	ErrInfo   bson.Raw
+	N         int32
+	NModified *int32
+	Upserted  *struct {
+		ID interface{} `bson:"_id"`
+	}
+}
+
+func (cur *cursorInfo) extractError() *WriteError {
+	if cur.Ok {
+		return nil
+	}
+	err := &WriteError{
+		Index:   int(cur.Idx),
+		Details: cur.ErrInfo,
+	}
+	if cur.Code != nil {
+		err.Code = int(*cur.Code)
+	}
+	if cur.Errmsg != nil {
+		err.Message = *cur.Errmsg
+	}
+	return err
+}
+
+type modelBatches struct {
+	session *session.Client
+	client  *Client
+
+	ordered bool
+	models  []clientWriteModel
+
+	offset int
+
+	retryMode      driver.RetryMode // RetryNone by default
+	cursorHandlers []func(*cursorInfo, bson.Raw) bool
+	newIDMap       map[int]interface{}
+
+	result             *ClientBulkWriteResult
+	writeConcernErrors []WriteConcernError
+	writeErrors        map[int]WriteError
+}
+
+func (mb *modelBatches) IsOrdered() *bool {
+	return &mb.ordered
+}
+
+func (mb *modelBatches) AdvanceBatches(n int) {
+	mb.offset += n
+	if mb.offset > len(mb.models) {
+		mb.offset = len(mb.models)
+	}
+}
+
+func (mb *modelBatches) Size() int {
+	if mb.offset > len(mb.models) {
+		return 0
+	}
+	return len(mb.models) - mb.offset
+}
+
+func (mb *modelBatches) AppendBatchSequence(dst []byte, maxCount, maxDocSize, totalSize int) (int, []byte, error) {
+	fn := functionSet{
+		appendStart: func(dst []byte, identifier string) (int32, []byte) {
+			var idx int32
+			dst = wiremessage.AppendMsgSectionType(dst, wiremessage.DocumentSequence)
+			idx, dst = bsoncore.ReserveLength(dst)
+			dst = append(dst, identifier...)
```

Comment:
From the specifications:

> The first entry in each document has the name of the operation as its key and the index ini the `nsInfo` array of the namespace on which the operation should be performed as its value

When I do command monitoring for client bulk write with multiple pairs, I get the following:

```
2024/11/18 14:39:47 started: &{Command:{"bulkWrite": {"$numberInt":"1"},"errorsOnly": false,"ordered": true,"lsid": {"id": {"$binary":{"base64":"XTDtLVGhTx6MEIcFDhf0qw==","subType":"04"}}},"txnNumber": {"$numberLong":"1"},"$clusterTime": {"clusterTime": {"$timestamp":{"t":1731965987,"i":1}},"signature": {"hash": {"$binary":{"base64":"AAAAAAAAAAAAAAAAAAAAAAAAAAA=","subType":"00"}},"keyId": {"$numberLong":"0"}}},"$db": "admin","ops": [{"insert": {"$numberInt":"0"},"document": {"_id": {"$oid":"673bb423a86efe4126c4c585"},"x": {"$numberInt":"1"}}},{"insert": {"$numberInt":"0"},"document": {"_id": {"$oid":"673bb423a86efe4126c4c586"},"x": {"$numberInt":"2"}}}],"nsInfo": [{"ns": "db.coll"}]} DatabaseName:admin CommandName:bulkWrite RequestID:1 ConnectionID:localhost:27017[-4] ServerConnectionID:0x14000390250 ServiceID:<nil>}
```

Where the index value for each document in the sequence is `{"$numberInt":"0"}`. Shouldn't this be `{"$numberInt":"0"}`, then `{"$numberInt":"1"}`, etc?

Reply:
The same thing occurs with the other client bulk write operations.

Reply:
Did you mean the Int32 value of the operation name such as "insert", "update", or "delete"?
```
...
"ops":[
    {
        "insert":{
            "$numberInt":"0"
        },
...
```

It is "the index in the `nsInfo` array of the namespace on which the operation should be performed as its value".

The specs also require:
> When constructing the `nsInfo` array for a `bulkWrite` batch, drivers MUST only include the namespaces that are referenced in the `ops` array for that batch.

and:
> Drivers MUST NOT include duplicate namespaces in this list".

Therefore, if both operations perform on the same namespace, the `nsInfo` array should contain only one item, and both operation indices are 0, pointing to "db.coll".

Reply:
I see, thanks for the explanation!